File: scripts/apply-foia-migration.py
#!/usr/bin/env python3
"""Apply FOIA stats migration via Supabase Management API.

Usage:
    SUPABASE_ACCESS_TOKEN=<token> python3 scripts/apply-foia-migration.py

Get token from: https://supabase.com/dashboard/account/tokens
Or run: supabase login
"""

import json, os, sys, re, ssl, urllib.request

PROJECT_REF = 'dzhqolbhuqdcpngdayuq'
MIGRATION_FILE = os.path.join(os.path.dirname(__file__), '..', 'supabase', 'migrations', '20260310_foia_block_ticket_stats.sql')

# SSL context — always disable verification since this system has cert issues
CTX = ssl.create_default_context()
CTX.check_hostname = False
CTX.verify_mode = ssl.CERT_NONE


def get_access_token():
    """Find Supabase access token from env or CLI config."""
    if os.environ.get('SUPABASE_ACCESS_TOKEN'):
        return os.environ['SUPABASE_ACCESS_TOKEN']

    home = os.path.expanduser('~')
    for p in [
        os.path.join(home, '.config', 'supabase', 'access-token'),
        os.path.join(home, '.supabase', 'access-token'),
    ]:
        if os.path.exists(p):
            return open(p).read().strip()

    # Try to get from supabase CLI
    import subprocess
    try:
        # v2.23 stores in OS keyring, extract via internal command
        result = subprocess.run(
            ['supabase', 'projects', 'list', '--output', 'json'],
            capture_output=True, text=True, timeout=15
        )
        if result.returncode == 0:
            # CLI is authenticated — try to use strace to find the token
            # Or better: look in credential store
            pass
    except:
        pass

    print("ERROR: No Supabase access token found.")
    print("Set SUPABASE_ACCESS_TOKEN env var.")
    print("Get a token from: https://supabase.com/dashboard/account/tokens")
    sys.exit(1)


def run_query(token, sql):
    """Execute SQL via Supabase Management API."""
    body = json.dumps({"query": sql}).encode()
    req = urllib.request.Request(
        f'https://api.supabase.com/v1/projects/{PROJECT_REF}/database/query',
        data=body,
        headers={
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json',
        },
    )
    try:
        resp = urllib.request.urlopen(req, context=CTX, timeout=30)
        return json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        raise Exception(f'HTTP {e.code}: {body[:500]}')


def main():
    print('=== Apply FOIA Migration ===\n')

    token = get_access_token()
    print(f'Token found ({len(token)} chars)')

    sql = open(MIGRATION_FILE).read()
    print(f'Migration SQL: {len(sql)} chars')

    # Run the entire migration as one statement
    print('\nExecuting migration...')
    try:
        result = run_query(token, sql)
        print(f'Result: {json.dumps(result)[:500]}')
    except Exception as e:
        print(f'Full execution failed: {e}')
        print('\nTrying statement by statement...')

        # Split into statements
        stmts = re.split(r';\s*\n', sql)
        stmts = [s.strip() for s in stmts if s.strip() and not s.strip().startswith('--')]

        for i, stmt in enumerate(stmts):
            preview = stmt[:80].replace('\n', ' ')
            sys.stdout.write(f'  [{i+1}/{len(stmts)}] {preview}...')
            sys.stdout.flush()
            try:
                run_query(token, stmt + ';')
                print(' OK')
            except Exception as e2:
                print(f' ERROR: {str(e2)[:200]}')

    # Verify
    print('\nVerifying tables...')
    try:
        result = run_query(token, """
            SELECT table_name FROM information_schema.tables
            WHERE table_schema='public' AND table_name LIKE 'foia_%'
            ORDER BY table_name;
        """)
        for row in result:
            print(f'  ✓ {row.get("table_name", row)}')
    except Exception as e:
        print(f'  Verify error: {e}')

    print('\nVerifying RPC functions...')
    try:
        result = run_query(token, """
            SELECT routine_name FROM information_schema.routines
            WHERE routine_schema='public' AND routine_name LIKE 'get_%ticket%'
            ORDER BY routine_name;
        """)
        for row in result:
            print(f'  ✓ {row.get("routine_name", row)}')
    except Exception as e:
        print(f'  Verify error: {e}')

    print('\n=== Done ===')


if __name__ == '__main__':
    main()

File: scripts/apply-foia-migration.ts
/**
 * Apply FOIA stats migration via Supabase Management API
 *
 * The Supabase Management API allows running arbitrary SQL via:
 *   POST https://api.supabase.com/v1/projects/{ref}/database/query
 *   Header: Authorization: Bearer <access_token>
 *   Body: { "query": "SQL here" }
 *
 * Alternatively, uses psql if DATABASE_URL is available.
 *
 * Usage:
 *   SUPABASE_ACCESS_TOKEN=<token> npx ts-node scripts/apply-foia-migration.ts
 *   (Get token from: supabase login, or https://supabase.com/dashboard/account/tokens)
 */

import * as fs from 'fs';
import * as path from 'path';
import * as https from 'https';

const PROJECT_REF = 'dzhqolbhuqdcpngdayuq';
const MIGRATION_FILE = path.join(__dirname, '..', 'supabase', 'migrations', '20260310_foia_block_ticket_stats.sql');

function getAccessToken(): string {
  // Check env
  if (process.env.SUPABASE_ACCESS_TOKEN) return process.env.SUPABASE_ACCESS_TOKEN;

  // Check common locations
  const homeDir = process.env.HOME || '';
  const paths = [
    path.join(homeDir, '.config', 'supabase', 'access-token'),
    path.join(homeDir, '.supabase', 'access-token'),
  ];

  for (const p of paths) {
    if (fs.existsSync(p)) {
      return fs.readFileSync(p, 'utf-8').trim();
    }
  }

  throw new Error(
    'No Supabase access token found.\n' +
    'Set SUPABASE_ACCESS_TOKEN env var or run: supabase login\n' +
    'Or get a token from: https://supabase.com/dashboard/account/tokens'
  );
}

function runQuery(token: string, sql: string): Promise<any> {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({ query: sql });
    const req = https.request({
      hostname: 'api.supabase.com',
      path: `/v1/projects/${PROJECT_REF}/database/query`,
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
      },
    }, (res) => {
      let data = '';
      res.on('data', (chunk) => data += chunk);
      res.on('end', () => {
        if (res.statusCode === 200 || res.statusCode === 201) {
          try { resolve(JSON.parse(data)); } catch { resolve(data); }
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    });
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

async function main() {
  console.log('=== Apply FOIA Migration ===\n');

  const token = getAccessToken();
  console.log(`Token found (${token.length} chars)`);

  const sql = fs.readFileSync(MIGRATION_FILE, 'utf-8');
  console.log(`Migration SQL: ${sql.length} chars from ${path.basename(MIGRATION_FILE)}`);

  // Split SQL into statements and run them individually
  // (some statements like CREATE TABLE + CREATE INDEX can fail independently)
  const statements = sql
    .split(/;\s*$/m)
    .map(s => s.trim())
    .filter(s => s.length > 0 && !s.startsWith('--'));

  console.log(`\n${statements.length} SQL statements to execute\n`);

  for (let i = 0; i < statements.length; i++) {
    const stmt = statements[i];
    const preview = stmt.substring(0, 80).replace(/\n/g, ' ');
    process.stdout.write(`  [${i + 1}/${statements.length}] ${preview}...`);

    try {
      await runQuery(token, stmt + ';');
      console.log(' OK');
    } catch (err: any) {
      console.log(` ERROR: ${err.message?.substring(0, 200)}`);
    }
  }

  console.log('\n=== Migration Complete ===');

  // Verify tables exist
  console.log('\nVerifying tables...');
  try {
    const result = await runQuery(token, `
      SELECT table_name
      FROM information_schema.tables
      WHERE table_schema = 'public'
        AND table_name LIKE 'foia_%'
      ORDER BY table_name;
    `);
    console.log('FOIA tables:', JSON.stringify(result));
  } catch (err: any) {
    console.error('Verify error:', err.message);
  }

  // Verify RPC functions
  try {
    const result = await runQuery(token, `
      SELECT routine_name
      FROM information_schema.routines
      WHERE routine_schema = 'public'
        AND routine_name LIKE 'get_%ticket%'
      ORDER BY routine_name;
    `);
    console.log('RPC functions:', JSON.stringify(result));
  } catch (err: any) {
    console.error('RPC verify error:', err.message);
  }
}

main().catch(console.error);
